---
title: "Using biomaRt to lookup gene symbols from RefSNP rs# IDs"
author: "Bill White"
date: "Monday, June 19, 2017"
output: html_document
---

# Access biomaRt to lookup genes from SNP refsnp IDs, such as GWAS "hits" from ranking methods

```{r eval=FALSE}
test-biomart.Rmd - Bill White - 6/18/17
R version 3.3.3 (2017-03-06)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 8 (jessie)
```

```{r}
library(knitr)
library(biomaRt) 
```

# Load the snp IDs from file

```{r}
snps.file <- "top-snps.txt"
the.snps <- read.table(snps.file, header=FALSE, stringsAsFactors=FALSE)[, 1]
kable(the.snps)
```

# Lookup the ensembl IDs

## Load the SNP info database and query the rs numbers

```{r}
snp2ensembl.biomart<- useMart("ENSEMBL_MART_SNP", dataset="hsapiens_snp")
snp.id.info <- getBM(c("refsnp_id", "ensembl_gene_stable_id", "ensembl_transcript_stable_id",
                       "chr_name", "chrom_start", "chrom_end"),
                     filters="snp_filter",
                     values=the.snps,
                     uniqueRows=FALSE,
                     mart=snp2ensembl.biomart)
map.ensembl.clean <- snp.id.info[snp.id.info$ensembl_gene_stable_id != "", ]
kable(map.ensembl.clean)
```

# Lookup the ensembl ID and get the gene symbol

## Load the gene info database and query the ensembl genes

```{r}
ensembl2gene.biomart <- useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
snp.gene.info <- getBM(attributes=c('ensembl_gene_id', 'hgnc_symbol', 'description'), 
                       filters='ensembl_transcript_id', 
                       values=map.ensembl.clean$ensembl_transcript_stable_id,
                       uniqueRows=TRUE,
                       mart=ensembl2gene.biomart)
kable(snp.gene.info)
```

# Format the results

```{r}
snp.genes.annot <- cbind(map.ensembl.clean[match(snp.gene.info$ensembl_gene_id, 
                                                 map.ensembl.clean$ensembl_gene_stable_id), 1], 
                         snp.gene.info)
colnames(snp.genes.annot) <- c("RefSNP", "Ensembl", "Gene", "Description")
# remove snps with no associated gene symbol
snp.genes.annot <- snp.genes.annot[snp.genes.annot$Gene != "", ]
# sort data frame returned by snp then gene ID
kable(snp.genes.annot[order(snp.genes.annot$RefSNP, snp.genes.annot$Ensembl), ])
```
